File: vibeshield-ts/src/middleware/safe-exec.ts
import { spawn } from 'child_process';
import { sanitizeShellInput, validateSafeCommand } from '../core/command-sanitizer.js';

/**
 * Result structure of a safe command execution.
 */
export interface ExecResult {
  stdout: string;
  stderr: string;
  code: number | null;
}

/**
 * Safely executes a command by validating the binary and sanitizing input arguments.
 * Forces shell: false to avoid command injection via shell execution features.
 *
 * @param command - The binary/command to run (e.g. 'ping', 'dig').
 * @param args - Array of command-line arguments.
 * @param options - Additional options to pass to spawn.
 * @param customAllowedList - Optional custom list of allowed commands.
 * @returns A promise resolving to the execution result.
 */
export function safeExec(
  command: string, 
  args: string[], 
  options: any = {}, 
  customAllowedList?: string[]
): Promise<ExecResult> {
  return new Promise((resolve, reject) => {
    try {
      validateSafeCommand(command, args, customAllowedList);
      const sanitizedArgs = args.map(arg => sanitizeShellInput(arg));
      
      const spawnOpts = {
        ...options,
        shell: false // FORCE shell: false
      };
      
      const child = spawn(command, sanitizedArgs, spawnOpts);
      
      let stdout = '';
      let stderr = '';
      
      child.stdout?.on('data', (data) => {
        stdout += data.toString();
      });
      
      child.stderr?.on('data', (data) => {
        stderr += data.toString();
      });
      
      child.on('error', (err) => {
        reject(err);
      });
      
      child.on('close', (code) => {
        resolve({ stdout, stderr, code });
      });
    } catch (err) {
      reject(err);
    }
  });
}

File: vibeshield-python/src/middleware/safe_exec.py
import subprocess
import logging
from ..command_sanitizer import sanitize_shell_input, validate_safe_command

logger = logging.getLogger("VibeShield")

def safe_exec(command: str, args: list[str], options: dict = None, allowed_commands: list[str] = None) -> dict:
    """
    Safely executes a shell command with shell=False.

    Args:
        command: The command name to execute.
        args: List of argument strings.
        options: Dict of subprocess.run options.
        allowed_commands: Optional custom allowed commands whitelist.

    Returns:
        Dict with keys:
            - stdout (str)
            - stderr (str)
            - code (int)
    """
    validate_safe_command(command, args, allowed_commands)
    sanitized_args = [sanitize_shell_input(arg) for arg in args]
    
    opts = options.copy() if options else {}
    # Force shell=False and capture output
    opts['shell'] = False
    opts['capture_output'] = True
    opts['text'] = True
    
    try:
        run_args = [command] + sanitized_args
        result = subprocess.run(run_args, **opts)
        return {
            'stdout': result.stdout,
            'stderr': result.stderr,
            'code': result.returncode
        }
    except Exception as e:
        logger.error("[VibeShield] Command execution failed: %s", str(e))
        raise
